### DISCUSSION TITLE
Service Worker Side Rendering

### DISCUSSION THREAD

**Discussion by thepassle** [withastro/roadmap#188] [Proposal]
This weekend I worked on a bit of a crazy PoC to see if I can run Astro in a service worker, and it turns out that you totally can:

https://user-images.githubusercontent.com/17054057/166226157-47d07ae5-0c7b-42e6-bb5d-1062e63b03ac.mov

This is huge, because:
- Your Astro apps are now available offline, as well as installable as PWA's
- This dramatically reduces the function invocations of the provider your app is deployed on
- There currently is not a single framework that makes PWA-ifying MPA's as easy as this

This lead to a bunch of other thoughts on how to improve this, and try to see how far I can take this. Currently the implementation I hacked together for this is a bit hacky and lives here [thepassle/astro-service-worker](https://github.com/thepassle/astro-service-worker). It's currently implemented as an _adapter_. This is not ideal, because currently having only 1 adapter in your `astro.config.mjs` is supported. 

Looking at Astro's source code, I realized perhaps a better fit would be an _integration_, that adds a vite/rollup plugin that creates a virtual mode to bundle the service worker code. I have some code put together in [this gist](https://gist.github.com/thepassle/be9010c330d7524133a4cfcf0a1c2ea1) to make this happen, but there are a few missing pieces of the puzzles, which I'll attempt to describe here.

## Missing pieces/apis to make this happen

Ideally, this would be be an integration, used like so:

```js
import { defineConfig } from 'astro/config';
import netlify from '@astrojs/netlify';

export default defineConfig({
  adapter: netlify(),
  integrations: [serviceWorker()]
});
```

There are a few things missing to make this happen though. In order to create the service worker bundle, I need to have access to the SSR Manifest, which currently is not available in any of the integration hooks or vite/rollup hooks. I've created an issue for this [here](https://github.com/withastro/astro/issues/3251). 

Additionally, it'd be nice to be able to import the `manifestReplace` and `pagesVirtualModuleId`, but they are not available in `astro/core`'s package export map. I've currently copied them from Astro's source code, and hardcoded them in my code, but this is fragile, because if Astro changes them, my code will break.

Ideally, I'd also have access to the pageData in my integration/vite plugin, which is in `internals` via `eachPageData(internals)`, but `internals` is not exposed anywhere either. Having access to internals also gives greater control over which code should be **server-only** and **service-worker-only**.

## Taking things to the next level: Streaming, offline-capable Astro apps

Then, I was reminded of this get-together by Jeff Posnick and Luke Edwards: https://cloudflare.tv/event/6ZJ5mEjrgcnCtBXBsUtyqV and Jeff's demos of streaming service workers. 

Dreaming even further, it would be even more amazing to be able to stitch together stream responses in Astro components.

```astro
---
import Header from '../src/components/Header.astro';
import Sidemenu from '../src/components/Sidemenu.astro';
import Footer from '../src/components/Footer.astro';
import { streamable } from 'astro';
---
<html>
  <Header/>
  <Sidemenu/>
  {streamable(() => fetch('/content/foo.md').then(render))}
  <Footer/>
</html>
```

In a similar fashion to this [Workbox example](https://glitch.com/edit/#!/workbox-streams?path=sw.js%3A26%3A0):

```js
import { strategy } from 'workbox-streams';
import { registerRoute } from 'workbox-core';

const streamResponse = strategy([
  () => caches.match(HEADER_CACHE_KEY, {cacheName: CACHE_NAME}),
  () => `<nav>sidebar<ul><li><a href="/about">about</a></li></ul></nav>`,
  ({event}) => apiStrategy.handle({
    event: event,
    request: new Request('/content/foo.md'),
  }),
  () => caches.match(FOOTER_CACHE_KEY, {cacheName: CACHE_NAME}),
]);

registerRoute('/foo', streamResponse);
```

As [Alex Russell says](https://twitter.com/slightlylate/status/1520857711163834368?s=20&t=l08UhgLfeAR3apcz20BYZg):
> This is awesome because it means that you can now get the document starting to request your (SW cached) CSS, JS, and other "header" resources *in parallel* with SW startup *and* the network fetch. None of the steps serialise until content comes back.

Given that the `render` function is a tagged template literal which returns an `AstroComponent`, it seems like something like this should not be completely impossible in the future, but likely requires some changes in Astro.

```js
class AstroComponent {
  constructor(htmlParts, expressions) {
    this.htmlParts = htmlParts;
    this.expressions = expressions;
  }
  get [Symbol.toStringTag]() {
    return "AstroComponent";
  }
  *[Symbol.iterator]() {
    const { htmlParts, expressions } = this;
    for (let i = 0; i < htmlParts.length; i++) {
      const html = htmlParts[i];
      const expression = expressions[i];
      yield markHTMLString(html);
      yield _render(expression);
    }
  }
}
```

E.g.: if `expression.type === 'streamable'`, execute the callback which returns the stream, and stream it along to the browser.

**Comment by matthewp** [comment]
Big +1 from me. Would want to talk about implementation details though. Want to avoid exposing a lot of internal code publicly that prevents us from making changes (don't want to accidentally break integrations). 

Can you explain what it is you want to do with `manifestReplace`, `pagesVirtualModuleId` and `internals`?

---

Streaming seems like a separate issue to me. We are very close to having streaming work without needing any special user-code at all. Assuming that a service worker integration is creating an App instance like in SSR, that should *just work* once we finish off streaming.

**Reply by thepassle** [reply]
> Would want to talk about implementation details though. 

For sure, lots of initial thoughts in the OP post.

> Can you explain what it is you want to do with manifestReplace, pagesVirtualModuleId and internals?

### `manifestReplace`/`pagesVirtualModuleId`

To create the service worker bundle, I essentially mimic the vite-plugin-ssr, in the sense that I add a virtual module, and bundle that as a service worker. To bundle the service worker (much like the SSR build), I need the SSR Manifest to pass to my Astro `App` instance, for `app.match` and `app.render`.

To instantiate the Astro app, I also need the pages, see the following code:

```js
const virtualSwModule = `
import { start } from '${adapter.serverEntrypoint}';

// 🚨 Need the pages to pass to _manifest
import * as _main from '${pagesVirtualModuleId}';
import { deserializeManifest as _deserializeManifest } from 'astro/app';

// 🚨 Need the manifestReplace/SSR Manifest
const _manifest = Object.assign(_deserializeManifest('${manifestReplace}'), {
  pageMap: _main.pageMap,
  renderers: _main.renderers
});

const _args = ${adapter.args ? JSON.stringify(adapter.args) : '{}'};

// the start function calls new App() internally
start(_manifest, _args);
`
```

### `internals`

As for `internals`, there's different ways to think about this really. Initially, I wanted a way to distinguish between `.astro` pages and `.js` endpoints. Initially, I thought to exclude `.js` entrypoints inside of `pages/` from the service worker build, and essentially have `.js` files be server-only endpoints; like for example accessing a database or something like that. That way, you could have server-rendered pages on initial visit, service-worker-rendered pages on subsequent visits, and server-only endpoints for fetching dynamic data.

I'm open for thoughts on that, however.

Initially, I thought the only way to achieve this ('this' being: excluding endpoint route handlers from the SW build), was to have access to `internals`, and then read `eachPageData`, filter out `type: 'endpoint'` and exclude those modules from the service worker build. However, I think you can also just check the file path in rollups `resolveId`/`load` hooks, so I'm not sure this is actually needed anymore.

> Streaming seems like a separate issue to me. We are very close to having streaming work without needing any special user-code at all. Assuming that a service worker integration is creating an App instance like in SSR, that should just work once we finish off streaming.

That sounds good to me, itd be nice if no extra work would be needed 👍 It'd be nice to make sure we could achieve the pattern I described in my initial post.

**Reply by matthewp** [reply]
Cool, so it sounds like what you really need is access to the SSR manifest so that you can inject it into your bundle, right? You don't really need manifestReplace/pagesVirtualModuleId, you can have your own replacement string and inject the manifest there, or am I wrong there?

If indeed the manifest is the only thing needed, I think having that be provided through an integration hook might be the way to go. You could listen for that hook, save the manifest, and then inject it into your own bundle before files are written / finished.  Let me know how this sounds.

**Reply by matthewp** [reply]
Hm, Now I'm wondering if I'm overthinking it. If you had `@@ASTRO_MANIFEST_REPLACE@@` in your code and we just replaced all occurrences of that in our SSR plugin, we wouldn't even need another hook and all of this ceremony.

**Reply by thepassle** [reply]
> You don't really need manifestReplace/pagesVirtualModuleId, you can have your own replacement string and inject the manifest there, or am I wrong there?

I do need `pagesVirtualModuleId` because I need to import the pages in my service worker.

As for which solution to go with; im also a bit unsure which is the best way to go about it. Replacing all occurences of `@@ASTRO_MANIFEST_REPLACE@@` seems like a little more 'magical' behavior compared to having a build hook that gets passed the manifest, and I could also imagine there may be other usecases to have access to the SSR manifest?

**Reply by matthewp** [reply]
Sorry for flip-flopping but I think I'm back on the hook idea. I worry that a global string replacement will break an extreme corner case like someone that wants to display Astro source code on a web page or something. Better to isolate the use-case in a hook, I think.

**Reply by thepassle** [reply]
Hook sounds fine to me 👍 What about `pagesVirtualModuleId` though?

**Reply by matthewp** [reply]
Yeah, we can provide that, I understand why you need it now. We might want to move that to `astro/app` (and give it a more generic name like `astroPagesRootId`) since that's an SSR implementation detail.

**Reply by thepassle** [reply]
Alright so I think we're in agreement on:
- Adding a new integration hook that receives the SSR manifest (what should the name of the hook be?)
- Adding a `astroPagesRootId` to `astro/app` and exporting it from there

If that all sounds good and if we have a name for the integration hook I can get started on a PR

**Reply by matthewp** [reply]
Probably `'astro:build:ssr'`, follows the other names decently I think.

Thanks for volunteering on this! Adding one fixture/test file in `packages/astro/test/` would be needed too.

**Reply by thepassle** [reply]
I've created a PR here: https://github.com/withastro/astro/pull/3262

Additionally, I'd still like to be able to exclude certain pages from the Service Worker build. This would allow for creating server-only code. It could be configurable like this:

```js
import { defineConfig } from 'astro/config';
import netlify from '@astrojs/netlify';
import serviceWorker from 'astro-service-worker';

export default defineConfig({
  adapter: netlify(),
  integrations: [
    serviceWorker({
      // excludes `src/pages/networkonly.astro` from the service worker build because it contains nodejs/server-only code
      networkOnly: ['/networkonly'],
    }),
  ]
});
```

Currently this is not possible, because pages are being created in `vite-plugin-pages`, and there is no way to control which pages get output in the final build, you just get all of them. Maybe we can add a configuration option to allow some more control over the vite-plugin-pages plugin? I'd be happy to make another PR for this, but we'd need to decide on the configuration option.

**Reply by thepassle** [reply]
Considering that the `buildConfig` already gets passed to the `vite-plugin-pages`, maybe we can add a `excludePages: ['/networkonly']` to the `buildConfig`? That would require a minimal change to the plugin code:

```diff
    load(id) {
      if (id === resolvedVirtualModuleId) {
        let importMap = "";
        let imports = [];
        let i = 0;
        for (const pageData of eachPageData(internals)) {
+         if(opts.buildConfig?.excludePages?.includes(pageData.route.pathname)) continue;
          const variable = `_page${i}`;
          imports.push(`import * as ${variable} from '${pageData.moduleSpecifier}';`);
          importMap += `['${pageData.component}', ${variable}],`;
          i++;
        }
        i = 0;
        let rendererItems = "";
        for (const renderer of opts.astroConfig._ctx.renderers) {
          const variable = `_renderer${i}`;
          imports.unshift(`import ${variable} from '${renderer.serverEntrypoint}';`);
          rendererItems += `Object.assign(${JSON.stringify(renderer)}, { ssr: ${variable} }),`;
          i++;
        }
        const def = `${imports.join("\n")}

export const pageMap = new Map([${importMap}]);
export const renderers = [${rendererItems}];`;
        return def;
      }
    }
```

Which could then be used in an integration like:

```js
'astro:build:start': async ({ buildConfig }) => {
  buildConfig.excludePages = ['/networkonly', 'etc'];
},
```

Or maybe some kind of `pageFilter` callback?
```diff
for (const pageData of eachPageData(internals)) {
+  if(opts.buildConfig?.pageFilter(pageData)) continue;
}
```

usage:
```js
'astro:build:start': async ({ buildConfig }) => {
  buildConfig.pageFilter = (pageData) => opts?.networkOnly?.includes(pageData.route.pathname);
},
```

**Reply by matthewp** [reply]
There isn't a separate build for service workers, right? It shares the same build if it's an integration. So excluding a page would expect it from being generated as part of the SSR build.

**Reply by thepassle** [reply]
Ah crap, youre right. Hrm. In that case I have to mimic the pages-plugin somehow, but I dont have access to the `eachPageData(internals)` 😕

do you have any other idea on how I could achieve this? Maybe the `internals.pagesByComponent.values()` can be exposed somewhere? 🤔 

The only way I see is by passing `pageData` to the `'astro:build:setup'` hook, e.g.: `await runHookBuildSetup({ pageData, config: astroConfig, vite: viteBuildConfig, target: "server" });`

Or maybe on `astroConfig._ctx.pageData`, since it also has other information like `renderers`, `adapter`, `scripts`, it seems in the same league, however, I think the pageData gets created after the `'astro:config:done'` hook, so I'm not sure if you'd have access to it 'in time' (before the rollup build)

**Reply by matthewp** [reply]
If what you are trying to achieve is a smaller bundle size by excluding parts you don't need in the service worker, I think the only real solution there would be for us to support multiple adapters; and therefore multiple builds.

But if you're goal is just to avoid *calling* API routes then I would think your SW code could handle that. When you call `app.match(request)` it returns a Route object which tells you if it's an API route. You can simple not call `app.render(request)` on those.

**Reply by thepassle** [reply]
What I'm trying to achieve is having a way of having server-only route handlers. Like, for example maybe my app needs to access a database, it'd be great to have my database connection/endpoint code in the same codebase. However, that endpoint will likely use dependencies that wont work in the browser/service worker, so I'd need to exclude those from the service worker bundle.

> But if you're goal is just to avoid calling API routes then I would think your SW code could handle that. When you call app.match(request) it returns a Route object which tells you if it's an API route. You can simple not call app.render(request) on those.

Yeah thats the implementation I have currently, but that would mean I'd have to deploy a separate app for doing server-only stuff.

**Reply by matthewp** [reply]
I get what you mean, that does seem like we probably want a separate bundle then. The same way that client / server get separate builds. So I think allowing an array of adapters is the answer here, where each one gets a separate build run on it, with some ability to control which routes are rendered like your suggestion here.

**Reply by thepassle** [reply]
But even if you make it a separate bundle, you still need access to the `pageData` to construct the imports

**Reply by thepassle** [reply]
Ah, sorry, I see what you're saying. Yeah, with multiple adapters that would be doable, but then you do need something like the `pageFilter` api I proposed above. 

Looking into the code, it seems like allowing an array of adapters is a non-trivial change though

**Reply by matthewp** [reply]
Based on the discussion in the API call, I think the next approach to take is to make your integration push in a top-level entrypoint.  You can see where the SSR plugin does the same thing here: https://github.com/withastro/astro/blob/main/packages/astro/src/core/build/vite-plugin-ssr.ts#L29-L31

Then I think you will need the pageData object, like you said. We might need another new hook to provide that at the time when you need it.

**Comment by thepassle** [comment]
Sidestepping the other thread a little bit for this discussion:

> Streaming seems like a separate issue to me. We are very close to having streaming work without needing any special user-code at all. Assuming that a service worker integration is creating an App instance like in SSR, that should just work once we finish off streaming.

Is there anywhere I can see this in action/progress? This already sounds great, but I'd love to make sure we'd be able to do something similar to the workbox example:

```js
import { strategy } from 'workbox-streams';
import { registerRoute } from 'workbox-core';

const streamResponse = strategy([
  () => caches.match(HEADER_CACHE_KEY, {cacheName: CACHE_NAME}),
  () => `<nav>sidebar<ul><li><a href="/about">about</a></li></ul></nav>`,
  ({event}) => apiStrategy.handle({
    event: event,
    request: new Request('/content/foo.md'),
  }),
  () => caches.match(FOOTER_CACHE_KEY, {cacheName: CACHE_NAME}),
]);

registerRoute('/foo', streamResponse);
```

**Reply by matthewp** [reply]
The way streaming in Astro would work is that each component blocks only itself. So if you had a component that did an API request in frontmatter, that would block rendering the template. But any other components that come before would not. So for example:

```astro
---
---
<Nav />
<FooContent />
```

Nav would be sent right away but FooContent would wait on API requests/etc. Does that answer your concerns?

----

There's possibly a little bit of work to get streaming, but not a lot. We need to get it prioritized, probably via an RFC of its own.

**Reply by thepassle** [reply]
What I'd really like to avoid is what Jake Archibald describes [in this blog](https://jakearchibald.com/2016/streams-ftw/#:~:text=HTML%20renders%20as%20it%20downloads):

> HTML renders as it downloads, whether it's served straight from a server or via a service worker. But I'm fetching the content from the page using JavaScript, then writing it to innerHTML, bypassing the streaming parser. Because of this, the content has to be fully downloaded before it can be displayed, and that's where the two second regression comes from. The more content you're downloading, the more the lack of streaming hurts performance, and unfortunately for me, Wikipedia articles are pretty big (the Google article is 100k).

<img width="757" alt="image" src="https://user-images.githubusercontent.com/17054057/166409788-0457154f-74a8-4ebe-9c00-45c950e25f05.png">

I understand that it could be cases where `<FooContent/>` does a network request that it has to wait for to finish completely, but in some cases the response could be streamed as well. Considering fetch's response.body is a stream as well, it'd be amazing to be able to write the following as a first class citizen:

`/blog/[id].astro`:
```js
---
import Header from '../src/components/Header.astro';
import Sidemenu from '../src/components/Sidemenu.astro';
import Footer from '../src/components/Footer.astro';
---
<html>
  <Header/>
  <Sidemenu/>
  {fetch(`/blog/${id}.html`).catch(() => {
    return caches?.match?.('/404.html') || fetch('/404.html');
  })}
  <Footer/>
</html>
```

Where `<Header/>` and `<Sidemenu/>` are completely static and can be streamed immediately, and then the result of the `fetch` can be streamed as well; as opposed for waiting for the fetch to completely finish downloading, and only _then_ start streaming the result of the fetch. (The example that Jake Archibald gives in his article is: Wikipedia articles can be pretty large, itd be great to immediately start streaming the response, instead of waiting for the entire article to have downloaded and then start responding)

Or even something being able to do something like:
```js
---
import Header from '../src/components/Header.astro';

function* renderLongList() {
  yield "item 1";
  yield "item 2";
}
---
<html>
  <Header/>
  {renderLongList()}
</html>
```

**Reply by matthewp** [reply]
Right, I think you are asking for support for Promises and Iterables in the Astro expression syntax right? Makes sense to me to add that.

**Reply by thepassle** [reply]
Awesome 🙂

**Reply by thepassle** [reply]
Following up on this, I've been hacking on this a little bit more, and here's some results:

https://user-images.githubusercontent.com/17054057/168627429-b98c17ae-2ac7-46b1-95df-e86e92cffacc.mov

Added some artificial delays for developing purposes. I am getting some double logs in my console, so it seems like I may be doing some unnecessary work (that I havent debugged yet), but for an exploration im not unhappy with the results.

It also supports inlining iterables and readable streams in astro expressions, e.g.:

```js
<div>
{() => fetch('/blog.html').then(r => r.body)}
</div>
```

The only issue is that I'm currently completely omitting the headInjection. I think you could work around that by keeping some kind of buffer before streaming, and if the buffer includes `</head>`, do the headInjection there, and send that.

The approach I took is largely: yield all the things. I tweaked the `renderPage`, `render`, `_render`, `renderComponent` and `renderSlot` functions to yield instead, and then creating a new ReadableStream and sending that as the response. Supporting iterables and readable streams is largely the same as @natemoo-re 's approach in microstream here: https://github.com/natemoo-re/microstream/blob/main/lib/mod.ts#L186-L218

**Comment by amatiasq** [comment]
This looks like an amazing idea! Has this approach gone further in recent years?
